python/taichi/_kernels.py

```diff
@@ -308,3 +310,143 @@ def parallel_sort(keys, values=None):
+    offset_j = 16
+    n = warp.shfl_up_i32(warp.active_mask(), val, offset_j)
+    if lane_id >= offset_j:
+        val += n
+    return val
+
+
+@kernel
+def scan_add_inclusive(arr_in: template(), in_beg: i32, in_end: i32,
+                       sum_smem: template(), single_block: template(),
+                       inclusive_add: template(), barrier: template()):
+    WARP_SZ = 32
+    BLOCK_SZ = 128
+    loop_config(block_dim=128)
+    for i in range(in_beg, in_end):
+        val = arr_in[i]
+
+        thread_id = i % BLOCK_SZ
+        block_id = int((i - in_beg) // BLOCK_SZ)
+        lane_id = thread_id % WARP_SZ
+        warp_id = thread_id // WARP_SZ
+
+        val = inclusive_add(val)
+        barrier()
+
+        # Put warp scan results to smem
+        # TODO replace smem with real smem when available
+        if thread_id % WARP_SZ == WARP_SZ - 1:
+            sum_smem[block_id, warp_id] = val
+        barrier()
+
+        # Inter-warp scan, use the first thread in the first warp
+        if warp_id == 0 and lane_id == 0:
+            for k in range(1, BLOCK_SZ / WARP_SZ):
+                sum_smem[block_id, k] += sum_smem[block_id, k - 1]
+        barrier()
+
+        # Update data with warp sums
+        warp_sum = 0
+        if warp_id > 0:
+            warp_sum = sum_smem[block_id, warp_id - 1]
+        val += warp_sum
+        arr_in[i] = val
+
+        # Update partial sums except the final block
+        if not single_block and (thread_id == BLOCK_SZ - 1):
+            arr_in[in_end + block_id] = val
+
+
+@kernel
+def uniform_add(arr_in: template(), in_beg: i32, in_end: i32):
+    BLOCK_SZ = 128
+    loop_config(block_dim=128)
+    for i in range(in_beg + BLOCK_SZ, in_end):
+        block_id = int((i - in_beg) // BLOCK_SZ)
+        arr_in[i] += arr_in[in_end + block_id - 1]
+
+
+@kernel
+def blit_from_field_to_field(
+        dst: template(), src: template(), offset: i32, size: i32):
+    for i in range(size):
+        dst[i + offset] = src[i]
+
+
+# Parallel Prefix Sum (Scan)
+# Ref[0]: https://developer.download.nvidia.com/compute/cuda/1.1-Beta/x86_website/projects/scan/doc/scan.pdf
+# Ref[1]: https://github.com/NVIDIA/cuda-samples/blob/master/Samples/2_Concepts_and_Techniques/shfl_scan/shfl_scan.cu
+def prefix_sum_inclusive_inplace(input_arr, length):
+    BLOCK_SZ = 128
+    GRID_SZ = int((length + BLOCK_SZ - 1) / BLOCK_SZ)
+
+    # Buffer position and length
+    # This is a single buffer implementation for ease of aot usage
+    ele_num = length
+    ele_nums = [ele_num]
+    start_pos = 0
+    ele_nums_pos = [start_pos]
+
+    while ele_num > 1:
+        ele_num = int((ele_num + BLOCK_SZ - 1) / BLOCK_SZ)
+        ele_nums.append(ele_num)
+        start_pos += BLOCK_SZ * ele_num
+        ele_nums_pos.append(start_pos)
+
+    if input_arr.dtype != i32:
+        raise RuntimeError("Only ti.i32 type is supported for prefix sum.")
+
+    large_arr = field(i32, shape=start_pos)
+    smem = field(i32, shape=(int(GRID_SZ), 64))
```

Comment:
OOC, does this only work on field?

Reply:
I think Ndarray should also work. Perhaps we can relax these limitations in a follow-up PR? Including support for float types etc.

Reply:
I see, we need to be careful about these implicit fields.